core/src/main/java/org/springframework/security/oauth2/server/authorization/jose/JoseHeaders.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization.jose;
+
+import org.springframework.core.convert.TypeDescriptor;
+import org.springframework.security.oauth2.core.converter.ClaimConversionService;
+import org.springframework.security.oauth2.jose.jws.MacAlgorithm;
+import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
+import org.springframework.util.Assert;
+
+import java.net.URI;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * A representation of &quot;headers&quot; that may be contained
+ * in the JSON object JOSE Header of a JSON Web Signature (JWS).
+ *
+ * @author Anoop Garlapati
+ * @since 0.0.1
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7515#section-4">JOSE Header of JWS</a>
+ */
+public class JoseHeaders {
+
+	private final Map<String, Object> headers;
+
+	protected JoseHeaders(Map<String, Object> headers) {
+		Assert.notEmpty(headers, "headers cannot be empty");
+		this.headers = Collections.unmodifiableMap(new LinkedHashMap<>(headers));
+	}
+
+	public SignatureAlgorithm getSignatureAlgorithm() {
+		return SignatureAlgorithm.from(getHeaderAsString("alg"));
+	}
+
+	public MacAlgorithm getMacAlgorithm() {
+		return MacAlgorithm.from(getHeaderAsString("alg"));
+	}
+
+	public String getType() {
+		return getHeaderAsString("typ");
+	}
+
+	public Map<String, Object> getHeaders() {
+		return this.headers;
+	}
+
+	public String getHeaderAsString(String header) {
```

Comment:
I realize that these are following the `JwtClaimAccessor` pattern; however, I think that it's better to coerce the values to the correct type on write instead of converting them on each read. 

Further, I think the `JwtClaimAccessor` makes a bit more sense when we are reading a JWT that another library has deserialized and chosen the types for each claim. Since we are the ones serializing in this case, I'm not seeing a huge need for these converters.

Is there any reason that we can't just have one method for now:

```java
public <H> H getHeader(String name) {
    return (H) this.headers.get(name)
}
```

and add the others later on, if needed?

Reply:
I agree that since we are the ones serializing we might not need converters. I will simplify this in my next draft.